general/development/tools/behat.md

```diff
@@ -0,0 +1,721 @@
+
+```
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun1/behat/behat.yml \
+        --format=pretty --out=/tmp/pretty_1.txt \
+        --format=moodle_progress --out=std
+
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun2/behat/behat.yml \
+        --format=pretty --out=/tmp/pretty_2.txt \
+        --format=moodle_progress --out=std
+
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun3/behat/behat.yml \
+        --format=pretty --out=/tmp/pretty_3.txt \
+        --format=moodle_progress --out=std
+```
+
+#### Rerun failed scenarios
+
+With slow systems or parallel run you may experience see some random failures. These may happen when your system is too slow, when it is too fast, or where a page depends on external dependencies.
+
+To help with this it is possible to rerun any failed scenarios using the `--rerun` option to Behat.
+
+The following example runs Behat with the rerun option:
+
+```
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml \
+        --format=pretty --out=/tmp/pretty.txt \
+        --format=moodle_progress --out=std \
+        --rerun
+```
+
+If any single test fails then the command will return a non-zero exit code. Running the same command again will mean that only failed scenarios are run.
+
+For a parallel run it can be called as follows:
+
+```
+    $ admin/tool/behat/cli/run.php \
+        --replace="{runprocess}" \
+        --format=pretty --out=/tmp/pretty_{runprocess}.txt \
+        --format=moodle_progress --out=std \
+        --rerun
+```
+
+The Moodle behat runner also includes an `--auto-rerun` option which will automatically rerun failed scenarios exactly once.
+
+#### Running behat with specified theme
+
+Behat can be run with any installed theme, but it must be initialised with the `-a` or `--add-core-features-to-theme` option first.
+
+After configuring the theme can be specified using the `--suite` parameter.
+
+**Note:** The default theme in Moodle (boost) has the suite name `default`.
+
+```
+    // Initialise Behat for all themes:
+    $ php admin/tool/behat/cli/init.php --add-core-features-to-theme
+
+    // Run Behat against all initalised themes:
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml
+
+    // Run Behat against just the default theme when all themes were initialised:
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml --suite=default
+
+    // Run Behat against just the 'classic' theme when all themes were initialised:
+    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml --suite=classic
+```
```

Comment:
```suggestion
:::note 

The default theme in Moodle (boost) has the suite name `default`.

```
    // Initialise Behat for all themes:
    $ php admin/tool/behat/cli/init.php --add-core-features-to-theme

    // Run Behat against all initalised themes:
    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml

    // Run Behat against just the default theme when all themes were initialised:
    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml --suite=default

    // Run Behat against just the 'classic' theme when all themes were initialised:
    $ vendor/bin/behat --config /Users/nicols/Sites/moodles/sm/moodledata_behat/behatrun/behat/behat.yml --suite=classic
```

:::
```